src/util/KvAuth0.js

```diff
@@ -0,0 +1,135 @@
+// These symbols are unique, and therefore are private to this scope.
+// For more details, see https://medium.com/@davidrhyswhite/private-members-in-es6-db1ccd6128a5
+const assertBrowser = Symbol('assertBrowser');
+const authCallback = Symbol('authCallback');
+const loginPromise = Symbol('loginPromise');
+const sessionPromise = Symbol('sessionPromise');
+
+// Class to handle interacting with auth0 in the browser
+export default class KvAuth0 {
+	constructor({
+		accessToken = '',
+		audience,
+		clientID,
+		domain,
+		redirectUri,
+		scope,
+		user = null,
+	}) {
+		this.enabled = true;
+		this.user = user;
+		this.accessToken = accessToken;
+		this.isServer = typeof window === 'undefined';
+
+		if (!this.isServer) {
+			// delay loading of webauth until we've confirmed we're in the browser,
+			// as it complains about not having window
+			this.webAuth = import('auth0-js').then(({ WebAuth }) => {
+				return new WebAuth({
+					audience,
+					clientID,
+					domain,
+					redirectUri,
+					responseType: 'token id_token',
+					scope,
+				});
+			});
+		}
+	}
+
+	// Default handler for the authentication result
+	[authCallback](err, result) {
+		return new Promise((resolve, reject) => {
+			if (err) {
+				switch (err.error) {
+					case 'login_required':
+					case 'consent_required':
+					case 'interaction_required':
+						// These errors require interaction beyond what can be provided by webauth,
+						// so resolve quietly without authentication
+						this.user = null;
+						this.accessToken = '';
+						console.log(`Auth session not started (${err.error_description})`);
+						resolve();
+						break;
+					default:
+						// Everything else, actually throw an error
+						console.error(err);
+						reject(err);
+				}
+			} else {
+				// Successful authentication
+				this.user = result.idTokenPayload;
+				this.accessToken = result.accessToken;
+				// handle expiration?
+				resolve(result);
+			}
+		});
+	}
+
+	// Throw an error if this gets called while not in a browser
```

Comment:
In what situation would a user be logging in and not be in a browser? Is this attempting to prevent command line hacking?

Reply:
The is to keep devs from accidentally calling this from server-side code

Reply:
Thanks @emuvente